include/dg/llvm/ValueRelations/StructureAnalyzer.hpp

```diff
@@ -0,0 +1,782 @@
+                        result.emplace_back(llvm::ConstantInt::get(c1->getType(), newCount), size);
+                        break;
+
+                    default:
+                        assert (0 && "unreachable");
+                }
+                // if we found two-constant operation, non of them can be binary
+                // operator to further unwrap
+                break;
+            }
+
+            // TODO use more info here
+            if (! c1 && ! c2) break;
+
+            // else one of c1, c2 is constant and the other is variable
+            const llvm::Value* param = nullptr;
+            if (c2) { c1 = c2; param = op->getOperand(0); }
+            else param = op->getOperand(1);
+
+            // now c1 is constant and param is variable
+            assert (c1 && param);
+
+            switch(op->getOpcode()) {
+
+                case llvm::Instruction::Add:
+                    result.emplace_back(param, size);
+                    break;
+
+                case llvm::Instruction::Mul:
+                    result.emplace_back(param, size * c1->getZExtValue());
+                    break;
+
+                default:
+                    assert (0 && "unreachable");
+            }
+            currentView = result.back();
+            // reachable only in this last c1 && param case
+        }
+        return result;
+    }
+
+#ifndef NDEBUG
+    void ddump() const {
+        std::cerr << "Allocated area:" << std::endl;
+        std::cerr << "    ptr " << debug::getValName(ptr) << std::endl;
+        std::cerr << "    count " << debug::getValName(originalSizeView.elementCount) << std::endl;
+        std::cerr << "    size " << originalSizeView.elementSize << std::endl;
+        std::cerr << std::endl;
+    }
+#endif
+};
+
+struct CallRelation {
+    std::vector<std::pair<const llvm::Value*, const llvm::Value*>> equalPairs;
+    VRLocation* callSite = nullptr;
+};
+
+class StructureAnalyzer {
+
+
+    const llvm::Module& module;
+    const std::map<const llvm::Instruction*, VRLocation*>& locationMapping;
+    const std::map<const llvm::BasicBlock*, std::unique_ptr<VRBBlock>>& blockMapping;
+
+    // holds vector of instructions, which are processed on any path back to given location
+    // is computed only for locations with more than one predecessor
+    std::map<VRLocation*, std::vector<const llvm::Instruction*>> inloopValues;
+
+    // holds vector of values, which are defined at given location
+    std::map<VRLocation*, std::set<const llvm::Value*>> defined;
+
+    const std::vector<unsigned> collected = { llvm::Instruction::Add,
+                                              llvm::Instruction::Sub,
+                                              llvm::Instruction::Mul };
+    std::map<unsigned, std::set<const llvm::Instruction*>> instructionSets;
+
+    std::vector<AllocatedArea> allocatedAreas;
+    const std::vector<std::string> handledAllocationFunctions = { "malloc",
```

Comment:
There is `AllocationFunction` enum in `dg/include/dg/AnalysisOptions.h` and related methods for this functionality.

Reply:
Pěkné, díky! Hm, v AnalysisOptions by se mi ale hodilo ještě isStackAlocationFunction a isHeapAlocationFunction, můžu je přidat? Případně je nějaký důvod, proč ty metody nejsou statické?

Reply:
There is no reason for the methods not to be static, you can patch that ;) Adding `isHeapAllocationFunction` is a good idea. Just note that `alloca` does not necessarily allocate memory on the stack (it is implementation-dependent). However, it behaves like allocating memory on the stack, so we can probably ignore this fact.

Reply:
I looked what it would take to make the methods static and I encountered two problems: the first is that it seems a non-integer static member cannot be initiated as a part of class definition and when I initialize it outside, it causes linking error (the by-the-book solution would be to initiate it in a cpp file).
Also there are several classes that inherit from this one and I did not want to risk breaking everything by interfering with too much code. I can look into it separately, if you thought it would be useful.